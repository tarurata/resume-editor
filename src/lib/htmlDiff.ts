/**
 * HTML Diff Utility
 * Provides safe HTML diffing with proper sanitization and token-level comparison
 */

export interface DiffToken {
    type: 'unchanged' | 'added' | 'removed'
    content: string
    isHtml?: boolean
}

export interface DiffOptions {
    ignoreWhitespace?: boolean
    ignoreCase?: boolean
    maxContext?: number
}

/**
 * Sanitizes HTML content to prevent XSS attacks
 */
export function sanitizeHtml(html: string): string {
    // Create a temporary div to parse HTML
    const temp = document.createElement('div')
    temp.textContent = html
    return temp.innerHTML
}

/**
 * Tokenizes HTML content into meaningful units
 */
export function tokenizeHtml(html: string): string[] {
    // Split by HTML tags and text content
    const tokens: string[] = []
    let current = html
    let lastIndex = 0

    // Match HTML tags, text content, and whitespace
    const regex = /<[^>]*>|[\s]+|[^\s<]+/g
    let match

    while ((match = regex.exec(current)) !== null) {
        if (match.index > lastIndex) {
            // Add any text between matches
            const text = current.slice(lastIndex, match.index)
            if (text.trim()) {
                tokens.push(text)
            }
        }

        tokens.push(match[0])
        lastIndex = match.index + match[0].length
    }

    // Add remaining text
    if (lastIndex < current.length) {
        const text = current.slice(lastIndex)
        if (text.trim()) {
            tokens.push(text)
        }
    }

    return tokens
}

/**
 * Normalizes whitespace in text content
 */
export function normalizeWhitespace(text: string): string {
    return text.replace(/\s+/g, ' ').trim()
}

/**
 * Checks if a token is HTML
 */
export function isHtmlToken(token: string): boolean {
    return token.startsWith('<') && token.endsWith('>')
}

/**
 * Checks if a token is whitespace
 */
export function isWhitespaceToken(token: string): boolean {
    return /^\s+$/.test(token)
}

/**
 * Creates a word-level diff for text content
 */
export function createTextDiff(original: string, current: string, options: DiffOptions = {}): DiffToken[] {
    const { ignoreWhitespace = false, ignoreCase = false } = options

    let origText = original
    let currText = current

    if (ignoreCase) {
        origText = origText.toLowerCase()
        currText = currText.toLowerCase()
    }

    if (ignoreWhitespace) {
        origText = normalizeWhitespace(origText)
        currText = normalizeWhitespace(currText)
    }

    const originalWords = origText.split(/(\s+)/)
    const currentWords = currText.split(/(\s+)/)

    const diff: DiffToken[] = []
    let i = 0, j = 0

    while (i < originalWords.length || j < currentWords.length) {
        const origWord = originalWords[i] || ''
        const currWord = currentWords[j] || ''

        if (origWord === currWord) {
            diff.push({ type: 'unchanged', content: original.split(/(\s+)/)[i] || '' })
            i++
            j++
        } else if (origWord === '') {
            diff.push({ type: 'added', content: current.split(/(\s+)/)[j] || '' })
            j++
        } else if (currWord === '') {
            diff.push({ type: 'removed', content: original.split(/(\s+)/)[i] || '' })
            i++
        } else {
            // Look ahead for matches
            let found = false
            const maxLookahead = Math.min(5, currentWords.length - j)

            for (let k = 1; k <= maxLookahead; k++) {
                if (originalWords[i] === currentWords[j + k]) {
                    // Add all words between j and j+k as added
                    for (let l = j; l < j + k; l++) {
                        diff.push({ type: 'added', content: current.split(/(\s+)/)[l] || '' })
                    }
                    j += k
                    found = true
                    break
                }
            }

            if (!found) {
                diff.push({ type: 'removed', content: original.split(/(\s+)/)[i] || '' })
                diff.push({ type: 'added', content: current.split(/(\s+)/)[j] || '' })
                i++
                j++
            }
        }
    }

    return diff
}

/**
 * Creates an HTML diff with proper token-level comparison
 */
export function createHtmlDiff(original: string, current: string, options: DiffOptions = {}): DiffToken[] {
    const { ignoreWhitespace = false } = options

    // Tokenize both HTML strings
    const originalTokens = tokenizeHtml(original)
    const currentTokens = tokenizeHtml(current)

    const diff: DiffToken[] = []
    let i = 0, j = 0

    while (i < originalTokens.length || j < currentTokens.length) {
        const origToken = originalTokens[i] || ''
        const currToken = currentTokens[j] || ''

        if (origToken === currToken) {
            diff.push({
                type: 'unchanged',
                content: origToken,
                isHtml: isHtmlToken(origToken)
            })
            i++
            j++
        } else if (origToken === '') {
            diff.push({
                type: 'added',
                content: currToken,
                isHtml: isHtmlToken(currToken)
            })
            j++
        } else if (currToken === '') {
            diff.push({
                type: 'removed',
                content: origToken,
                isHtml: isHtmlToken(origToken)
            })
            i++
        } else if (isHtmlToken(origToken) && isHtmlToken(currToken)) {
            // Both are HTML tokens - compare as-is
            diff.push({ type: 'removed', content: origToken, isHtml: true })
            diff.push({ type: 'added', content: currToken, isHtml: true })
            i++
            j++
        } else if (isHtmlToken(origToken)) {
            // Original is HTML, current is text
            diff.push({ type: 'removed', content: origToken, isHtml: true })
            diff.push({ type: 'added', content: currToken, isHtml: false })
            i++
            j++
        } else if (isHtmlToken(currToken)) {
            // Current is HTML, original is text
            diff.push({ type: 'removed', content: origToken, isHtml: false })
            diff.push({ type: 'added', content: currToken, isHtml: true })
            i++
            j++
        } else {
            // Both are text tokens - use text diff algorithm
            const textDiff = createTextDiff(origToken, currToken, options)
            diff.push(...textDiff.map(token => ({ ...token, isHtml: false })))
            i++
            j++
        }
    }

    return diff
}

/**
 * Renders diff tokens as HTML with proper escaping
 */
export function renderDiffTokens(tokens: DiffToken[]): string {
    return tokens.map(token => {
        const escapedContent = token.isHtml ? token.content : sanitizeHtml(token.content)

        switch (token.type) {
            case 'added':
                return `<span class="diff-added">${escapedContent}</span>`
            case 'removed':
                return `<span class="diff-removed">${escapedContent}</span>`
            case 'unchanged':
            default:
                return escapedContent
        }
    }).join('')
}

/**
 * Main function to generate HTML diff
 */
export function generateHtmlDiff(original: string, current: string, options: DiffOptions = {}): string {
    const diff = createHtmlDiff(original, current, options)
    return renderDiffTokens(diff)
}
