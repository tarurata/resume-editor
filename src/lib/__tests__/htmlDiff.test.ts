/**
 * Snapshot tests for HTML diff functionality
 */

import { generateHtmlDiff, createHtmlDiff, DiffOptions } from '../htmlDiff'

describe('HTML Diff Snapshot Tests', () => {
    describe('generateHtmlDiff', () => {
        it('should generate diff for simple text changes', () => {
            const original = 'Hello world'
            const current = 'Hello beautiful world'
            const result = generateHtmlDiff(original, current)
            expect(result).toMatchSnapshot()
        })

        it('should generate diff for HTML content changes', () => {
            const original = '<p>Hello <strong>world</strong></p>'
            const current = '<p>Hello <em>beautiful</em> <strong>world</strong></p>'
            const result = generateHtmlDiff(original, current)
            expect(result).toMatchSnapshot()
        })

        it('should generate diff for list changes', () => {
            const original = '<ul><li>Item 1</li><li>Item 2</li></ul>'
            const current = '<ul><li>Item 1</li><li>New Item</li><li>Item 2</li></ul>'
            const result = generateHtmlDiff(original, current)
            expect(result).toMatchSnapshot()
        })

        it('should generate diff for complex HTML structure changes', () => {
            const original = '<div><h2>Experience</h2><p>Worked at Company A</p></div>'
            const current = '<div><h2>Professional Experience</h2><p>Worked at Company A and Company B</p></div>'
            const result = generateHtmlDiff(original, current)
            expect(result).toMatchSnapshot()
        })

        it('should handle whitespace changes when ignoreWhitespace is true', () => {
            const original = 'Hello    world'
            const current = 'Hello world'
            const options: DiffOptions = { ignoreWhitespace: true }
            const result = generateHtmlDiff(original, current, options)
            expect(result).toMatchSnapshot()
        })

        it('should show whitespace changes when ignoreWhitespace is false', () => {
            const original = 'Hello    world'
            const current = 'Hello world'
            const options: DiffOptions = { ignoreWhitespace: false }
            const result = generateHtmlDiff(original, current, options)
            expect(result).toMatchSnapshot()
        })

        it('should handle case sensitivity when ignoreCase is true', () => {
            const original = 'Hello World'
            const current = 'hello world'
            const options: DiffOptions = { ignoreCase: true }
            const result = generateHtmlDiff(original, current, options)
            expect(result).toMatchSnapshot()
        })

        it('should show case changes when ignoreCase is false', () => {
            const original = 'Hello World'
            const current = 'hello world'
            const options: DiffOptions = { ignoreCase: false }
            const result = generateHtmlDiff(original, current, options)
            expect(result).toMatchSnapshot()
        })

        it('should handle empty content', () => {
            const original = ''
            const current = '<p>New content</p>'
            const result = generateHtmlDiff(original, current)
            expect(result).toMatchSnapshot()
        })

        it('should handle identical content', () => {
            const original = '<p>Same content</p>'
            const current = '<p>Same content</p>'
            const result = generateHtmlDiff(original, current)
            expect(result).toMatchSnapshot()
        })

        it('should handle malformed HTML safely', () => {
            const original = '<p>Hello <unclosed>world</p>'
            const current = '<p>Hello <strong>world</strong></p>'
            const result = generateHtmlDiff(original, current)
            expect(result).toMatchSnapshot()
        })

        it('should handle script tags safely', () => {
            const original = '<p>Hello world</p>'
            const current = '<p>Hello <script>alert("xss")</script>world</p>'
            const result = generateHtmlDiff(original, current)
            expect(result).toMatchSnapshot()
        })
    })

    describe('createHtmlDiff', () => {
        it('should create token-level diff for HTML content', () => {
            const original = '<p>Hello <strong>world</strong></p>'
            const current = '<p>Hello <em>beautiful</em> <strong>world</strong></p>'
            const result = createHtmlDiff(original, current)
            expect(result).toMatchSnapshot()
        })

        it('should handle mixed HTML and text changes', () => {
            const original = 'Text before <span>HTML</span> text after'
            const current = 'Text before <div>HTML content</div> text after'
            const result = createHtmlDiff(original, current)
            expect(result).toMatchSnapshot()
        })

        it('should preserve HTML structure in unchanged parts', () => {
            const original = '<div><h1>Title</h1><p>Content</p></div>'
            const current = '<div><h1>New Title</h1><p>Content</p></div>'
            const result = createHtmlDiff(original, current)
            expect(result).toMatchSnapshot()
        })
    })
})
