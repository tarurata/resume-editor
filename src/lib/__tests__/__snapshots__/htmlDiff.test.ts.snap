// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`HTML Diff Snapshot Tests createHtmlDiff should create token-level diff for HTML content 1`] = `
[
  {
    "content": "<p>",
    "isHtml": true,
    "type": "unchanged",
  },
  {
    "content": "Hello",
    "isHtml": false,
    "type": "unchanged",
  },
  {
    "content": " ",
    "isHtml": false,
    "type": "unchanged",
  },
  {
    "content": "<strong>",
    "isHtml": true,
    "type": "removed",
  },
  {
    "content": "<em>",
    "isHtml": true,
    "type": "added",
  },
  {
    "content": "world",
    "isHtml": false,
    "type": "removed",
  },
  {
    "content": "beautiful",
    "isHtml": false,
    "type": "added",
  },
  {
    "content": "</strong>",
    "isHtml": true,
    "type": "removed",
  },
  {
    "content": "</em>",
    "isHtml": true,
    "type": "added",
  },
  {
    "content": "</p>",
    "isHtml": true,
    "type": "removed",
  },
  {
    "content": " ",
    "isHtml": false,
    "type": "added",
  },
  {
    "content": "<strong>",
    "isHtml": true,
    "type": "added",
  },
  {
    "content": "world",
    "isHtml": false,
    "type": "added",
  },
  {
    "content": "</strong>",
    "isHtml": true,
    "type": "added",
  },
  {
    "content": "</p>",
    "isHtml": true,
    "type": "added",
  },
]
`;

exports[`HTML Diff Snapshot Tests createHtmlDiff should handle mixed HTML and text changes 1`] = `
[
  {
    "content": "Text",
    "isHtml": false,
    "type": "unchanged",
  },
  {
    "content": " ",
    "isHtml": false,
    "type": "unchanged",
  },
  {
    "content": "before",
    "isHtml": false,
    "type": "unchanged",
  },
  {
    "content": " ",
    "isHtml": false,
    "type": "unchanged",
  },
  {
    "content": "<span>",
    "isHtml": true,
    "type": "removed",
  },
  {
    "content": "<div>",
    "isHtml": true,
    "type": "added",
  },
  {
    "content": "HTML",
    "isHtml": false,
    "type": "unchanged",
  },
  {
    "content": "</span>",
    "isHtml": true,
    "type": "removed",
  },
  {
    "content": " ",
    "isHtml": false,
    "type": "added",
  },
  {
    "content": "content",
    "isHtml": false,
    "type": "added",
  },
  {
    "content": "",
    "isHtml": false,
    "type": "unchanged",
  },
  {
    "content": " ",
    "isHtml": false,
    "type": "removed",
  },
  {
    "content": "",
    "isHtml": false,
    "type": "unchanged",
  },
  {
    "content": "text",
    "isHtml": false,
    "type": "removed",
  },
  {
    "content": "</div>",
    "isHtml": true,
    "type": "added",
  },
  {
    "content": " ",
    "isHtml": false,
    "type": "unchanged",
  },
  {
    "content": "after",
    "isHtml": false,
    "type": "removed",
  },
  {
    "content": "text",
    "isHtml": false,
    "type": "added",
  },
  {
    "content": " ",
    "isHtml": false,
    "type": "added",
  },
  {
    "content": "after",
    "isHtml": false,
    "type": "added",
  },
]
`;

exports[`HTML Diff Snapshot Tests createHtmlDiff should preserve HTML structure in unchanged parts 1`] = `
[
  {
    "content": "<div>",
    "isHtml": true,
    "type": "unchanged",
  },
  {
    "content": "<h1>",
    "isHtml": true,
    "type": "unchanged",
  },
  {
    "content": "Title",
    "isHtml": false,
    "type": "removed",
  },
  {
    "content": "New",
    "isHtml": false,
    "type": "added",
  },
  {
    "content": "</h1>",
    "isHtml": true,
    "type": "removed",
  },
  {
    "content": " ",
    "isHtml": false,
    "type": "added",
  },
  {
    "content": "<p>",
    "isHtml": true,
    "type": "removed",
  },
  {
    "content": "Title",
    "isHtml": false,
    "type": "added",
  },
  {
    "content": "Content",
    "isHtml": false,
    "type": "removed",
  },
  {
    "content": "</h1>",
    "isHtml": true,
    "type": "added",
  },
  {
    "content": "</p>",
    "isHtml": true,
    "type": "removed",
  },
  {
    "content": "<p>",
    "isHtml": true,
    "type": "added",
  },
  {
    "content": "</div>",
    "isHtml": true,
    "type": "removed",
  },
  {
    "content": "Content",
    "isHtml": false,
    "type": "added",
  },
  {
    "content": "</p>",
    "isHtml": true,
    "type": "added",
  },
  {
    "content": "</div>",
    "isHtml": true,
    "type": "added",
  },
]
`;

exports[`HTML Diff Snapshot Tests generateHtmlDiff should generate diff for HTML content changes 1`] = `"<p>Hello <span class="diff-removed"><strong></span><span class="diff-added"><em></span><span class="diff-removed">world</span><span class="diff-added">beautiful</span><span class="diff-removed"></strong></span><span class="diff-added"></em></span><span class="diff-removed"></p></span><span class="diff-added"> </span><span class="diff-added"><strong></span><span class="diff-added">world</span><span class="diff-added"></strong></span><span class="diff-added"></p></span>"`;

exports[`HTML Diff Snapshot Tests generateHtmlDiff should generate diff for complex HTML structure changes 1`] = `"<div><h2><span class="diff-removed">Experience</span><span class="diff-added">Professional</span><span class="diff-removed"></h2></span><span class="diff-added"> </span><span class="diff-removed"><p></span><span class="diff-added">Experience</span><span class="diff-removed">Worked</span><span class="diff-added"></h2></span><span class="diff-removed"> </span><span class="diff-added"><p></span><span class="diff-removed">at</span><span class="diff-added">Worked</span> <span class="diff-removed">Company</span><span class="diff-added">at</span> <span class="diff-removed">A</span><span class="diff-added">Company</span><span class="diff-removed"></p></span><span class="diff-added"> </span><span class="diff-removed"></div></span><span class="diff-added">A</span><span class="diff-added"> </span><span class="diff-added">and</span><span class="diff-added"> </span><span class="diff-added">Company</span><span class="diff-added"> </span><span class="diff-added">B</span><span class="diff-added"></p></span><span class="diff-added"></div></span>"`;

exports[`HTML Diff Snapshot Tests generateHtmlDiff should generate diff for list changes 1`] = `"<ul><li>Item 1</li><li><span class="diff-removed">Item</span><span class="diff-added">New</span> <span class="diff-removed">2</span><span class="diff-added">Item</span></li><span class="diff-removed"></ul></span><span class="diff-added"><li></span><span class="diff-added">Item</span><span class="diff-added"> </span><span class="diff-added">2</span><span class="diff-added"></li></span><span class="diff-added"></ul></span>"`;

exports[`HTML Diff Snapshot Tests generateHtmlDiff should generate diff for simple text changes 1`] = `"Hello <span class="diff-removed">world</span><span class="diff-added">beautiful</span><span class="diff-added"> </span><span class="diff-added">world</span>"`;

exports[`HTML Diff Snapshot Tests generateHtmlDiff should handle case sensitivity when ignoreCase is true 1`] = `"Hello World"`;

exports[`HTML Diff Snapshot Tests generateHtmlDiff should handle empty content 1`] = `"<span class="diff-added"><p></span><span class="diff-added">New</span><span class="diff-added"> </span><span class="diff-added">content</span><span class="diff-added"></p></span>"`;

exports[`HTML Diff Snapshot Tests generateHtmlDiff should handle identical content 1`] = `"<p>Same content</p>"`;

exports[`HTML Diff Snapshot Tests generateHtmlDiff should handle malformed HTML safely 1`] = `"<p>Hello <span class="diff-removed"><unclosed></span><span class="diff-added"><strong></span>world<span class="diff-removed"></p></span><span class="diff-added"></strong></span><span class="diff-added"></p></span>"`;

exports[`HTML Diff Snapshot Tests generateHtmlDiff should handle script tags safely 1`] = `"<p>Hello <span class="diff-removed">world</span><span class="diff-added"><script></span><span class="diff-removed"></p></span><span class="diff-added">alert("xss")</span><span class="diff-added"></script></span><span class="diff-added">world</span><span class="diff-added"></p></span>"`;

exports[`HTML Diff Snapshot Tests generateHtmlDiff should handle whitespace changes when ignoreWhitespace is true 1`] = `"Helloworld"`;

exports[`HTML Diff Snapshot Tests generateHtmlDiff should show case changes when ignoreCase is false 1`] = `"<span class="diff-removed">Hello</span><span class="diff-added">hello</span> <span class="diff-removed">World</span><span class="diff-added">world</span>"`;

exports[`HTML Diff Snapshot Tests generateHtmlDiff should show whitespace changes when ignoreWhitespace is false 1`] = `"Hello<span class="diff-removed">    </span><span class="diff-added"> </span>world"`;
